stdenv: fix build times

happysalada opened the pull request:
###### Motivation for this change

This is not a major concern, but there are two problems with the code
- it uses cat instead of built in file redirection
- There is a shellcheck warning on unclear array semantics

This can be tested with the small script
```
buildTimes=()
while read -r line; do
    echo "$line"
    buildTimes+=("$line")
done < <(times) > .times
echo "build time elapsed: " "${buildTimes[@]}"
```

@7c6f434c you've been kind enough to review some bash before.
@andersk you've been very knowledgeable about shell

No worries if you don't have time.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

veprbl commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -121,14 +121,17 @@ exitHandler() {
-        times > "$NIX_BUILD_TOP/.times"
-        local -a times=($(cat "$NIX_BUILD_TOP/.times"))
+        buildTimes=()
+        while read -r line; do
```
I wonder if this could work to give a bit more natural code ordering
```suggestion
        times | while read -r line; do
```

happysalada replied:
Yeah I tried that first, I didn't know as well, but there was a shellcheck warning, it taught me something.
Basically if you do a pipe in a loop it doesn't modify the variable in the outer context. (I'm sure you've figured it out, just putting it out for other people wondering).

andersk commented:
Using strings would be simpler than arrays here.

```bash
local times
times="$(times)"
printf '%s\n' "$times" > .times
echo "build time elapsed: ${times//$'\n'/ }"
```

(In cases where you need an array, the `mapfile` builtin is good for that: `mapfile -t timesArray <<<"$times"`.)

happysalada commented:
Thanks a lot for this!

What do you think about the following
```
times="$(times)"
read -r -d '' -a buildTimes <<<"$times"
echo "build time elapsed: user time for the shell               ${buildTimes[0]}"
echo "build time elapsed: system time for the shell             ${buildTimes[1]}"
echo "build time elapsed: user time for all child processes     ${buildTimes[2]}"
echo "build time elapsed: system time for all child processes   ${buildTimes[3]}"
```
That way we get rid of the comment at the same time ?

andersk commented:
I’m guessing we don’t actually need the `$NIX_BUILD_TOP/.times` file. `$NIX_BUILD_TOP` is a temporary directory; the `.times` file looks like it’s just a side effect of how the existing code was written, and there are no other references to it.

Not sure if anything cares about the format of the `build time elapsed:` line.

happysalada commented:
I was about to post the same thing. I don't think we need the file as well.

Regarding the format comment, are you referring to the spaces introduced to make it prettier ? I can remove those.

There is a comment in the code that says
```
        # Print the following statistics:
        # - user time for the shell
        # - system time for the shell
        # - user time for all child processes
        # - system time for all child processes
```

I was thinking that by adding the text in the after the `:` then we could get rid of it and it would be clearer to the user. I mostly added the empty spaces to make it easier to read.

andersk commented:
What I’m saying is there could be some script parsing the output of the `build time elapsed:` line, and would therefore expect it to be formatted like it is now:

```
build time elapsed:  0m0.086s 0m0.029s 0m37.204s 0m4.915s
```

I don’t know whether there *is* such a script.

happysalada commented:
I see, it makes sense, a quick grep didn't find anything relevant. (it's not a guarantee, but I think it makes me more confident).

happysalada committed:
stdenv: fix showBuildStats

- remove going through another file
- use builtin instead of external cat
- improve echo formatting

happysalada commented:
@andersk What do you think about removing the intermediate string variable `times` completely ?
(i've rebased and squashed, so I think the branch is ready for merging now)

andersk commented:
If there were an error from `times`, a standalone assignment `var="$(times)"` would catch it and abort under `set -e`, while a more complicated statement using `$(times)` or `<(times)` will not. That’s probably okay here since `times` is specified to never fail, though it’s a good general principle to keep in mind.

happysalada commented:
thank you! I completely agree with you point.
I see what you mean here. I'm a little on the fence for changing this, leaning on the leaving it like this side.
let me know if you feel like changing it.
